summary_workflows/quantification/match_with_gt.py

```diff
@@ -0,0 +1,187 @@
+import argparse
+
+parser = argparse.ArgumentParser()
+
+parser.add_argument('a', help='The BED file containing predictions. MUST be BED6 format.')
+parser.add_argument('b', help='The ground truth bed file. First 6 columns must be standard BED6, but can have additional columns appended.')
+parser.add_argument('window', help='Number of bases to append to each side of the predicted site.', type=int)
+#parser.add_argument('-o', help='output file directory') # not yet implemented!!
+args = parser.parse_args()
+
+
+f_GT = args.b
+f_PD = args.a
+window = args.window
+
+
+def bedtools_window(bed1, bed2, window):
+    """
+    Python wrapper for bedtools window.
+    """
+    
+    # make sure bedtools can be called in current env
+    assert which('bedtools') is not None, "bedtools not installed or not in PATH"
+
+    # run bedtools window, capture output
+    out = subprocess.run(['bedtools', 'window', '-sm',
+                          '-w', str(window), 
+                          '-a', bed1, 
+                          '-b', bed2], 
+                         capture_output=True, shell=False)
+
+    assert out.returncode==0, "bedtools window run failed, check input files"
+
+    # memory file-handle to pass output to pandas without writing to disk
+    out_handle = StringIO(out.stdout.decode())
+    out = pd.read_csv(out_handle, delimiter='\t', header=None, dtype={0: str})
+    
+    return(out)
+
+def overlap(coords1, coords2):
+    """
+    Calculates number of points overlapping when the input are two intervals as tuples.
+    Input must be coords1=(int, int),  coords2=(int, int).
+    """
+    return(len(np.intersect1d(np.arange(coords1[0], coords1[1]), np.arange(coords2[0], coords2[1]))))
+
+
+print('-------------------------------------------------')
+## sum expression levels for predicted sites that overlap with each other given the window size
+## assuming the prediction bed file has exactly 6 columns in the correct order, no header!
+
+# find sites that overlap within the predict set itself when the window is extended
+out = bedtools_window(f_PD, f_PD, window)
+
+# for logging
+len_dup = len(out)
+
+not_unq_mask = out.duplicated([0, 1, 2, 5], keep=False)
+
+MERGED=False
+if np.sum(not_unq_mask) > 0: # otherwise there were no overlapping sites
+    MERGED=True
+    
+    not_unq = out[not_unq_mask]
+
+    # remove non-unique sites
+    out.drop(out.index[not_unq_mask], inplace=True, axis=0)
+    out.drop([6, 7, 8, 9, 10, 11], inplace=True, axis=1)
+
+    # collapse non-unique sites, sum up expression, add back to bed file
+    unq = np.unique([(i, j, k, l) for i, j, k, l in zip(not_unq[0], not_unq[1], not_unq[2], not_unq[5])], axis=0)
+
+    for i in unq:
+        non_unq_context = not_unq[(not_unq[0]==i[0]) & (not_unq[1]==int(i[1])) & (not_unq[2]==int(i[2])) & (not_unq[5]==i[3])]
+
+        merged_start = min(list(non_unq_context[7]))
+        merged_end = max(list(non_unq_context[8]))
+        merged_exp = sum(non_unq_context[10])
```

Comment:
Here it seems that the expression of all ground truth sites that overlap with a predicted site is summed up. Is this intended?

Reply:
I'm not sure if I fully understand the code, but I think this part relates to summing up expression of all predicted PAS that are within the same window (`bedtools window` is run on predicted PAS both as input A and input B).

Reply:
Yes exactly! This was @mzavolan's suggestion in issue #64. Hope I understood that correctly.

Reply:
I see, it makes sense. But I'm wondering if it's necessary in this case when we have both a window for merging and for mapping to GT? Maybe it should be one or the other? Or at least maybe the windows should have different values, i.e. one small, one bigger?

Reply:
> Yes exactly! This was @mzavolan's suggestion in issue #64. Hope I understood that correctly.
> 
> It seems like it's the ground truth sites, but it's actually the predicted sites because bedtools window is called with the same input file twice. I'll make that more readable.

Ah, yes, it makes sense, missed that trying to resolve indices. The comments are there, but could be clearer.

Reply:
> I see, it makes sense. But I'm wondering if it's necessary in this case when we have both a window for merging and for mapping to GT? Maybe it should be one or the other? Or at least maybe the windows should have different values, i.e. one small, one bigger?

Yes, this relates to my comment about splitting predicted sites that match multiple ground truth sites. In principle, we trust the ground truth sites to be precise.

Reply:
We can have a different window for merging. What would you suggest? 

@mzavolan would you suggest splitting one predicted site that matches multiple ground truth sites into equal parts?